wireguard-broker/src/api/mio_client.rs

```diff
@@ -0,0 +1,231 @@
+            recv_state: RxState::RxSize(0),
+            recv_buf: [0u8; RECV_BUF_SIZE],
+            expected_state: RxState::RxSize(LEN_SIZE),
+        };
+        let inner = BrokerClient::new(io);
+        Self { inner }
+    }
+
+    pub fn poll(&mut self) -> anyhow::Result<Option<msgs::SetPskResult>> {
+        self.inner.io_mut().flush()?;
+
+        // This sucks
+        match self.inner.poll_response() {
+            Ok(res) => {
+                return Ok(res);
+            }
+            Err(BrokerClientPollResponseError::IoError(e)) => {
+                return Err(e);
+            }
+            Err(BrokerClientPollResponseError::InvalidMessage) => {
+                bail!("Invalid message");
+            }
+        };
+    }
+}
+
+impl WireGuardBroker for MioBrokerClient {
+    type Error = anyhow::Error;
+
+    fn set_psk(&mut self, iface: &str, peer_id: [u8; 32], psk: [u8; 32]) -> anyhow::Result<()> {
+        use BrokerClientSetPskError::*;
+        let e = self.inner.set_psk(iface, peer_id, psk);
+        match e {
+            Ok(()) => Ok(()),
+            Err(IoError(e)) => Err(e),
+            Err(IfaceOutOfBounds) => bail!("Interface name size is out of bounds."),
+            Err(MsgError) => bail!("Error with encoding/decoding message."),
+        }
+    }
+}
+
+impl BrokerClientIo for MioBrokerClientIo {
+    type SendError = anyhow::Error;
+    type RecvError = anyhow::Error;
+
+    fn send_msg(&mut self, buf: &[u8]) -> Result<(), Self::SendError> {
+        self.flush()?;
+        self.send_or_buffer(&(buf.len() as u64).to_le_bytes())?;
+        self.send_or_buffer(&buf)?;
+        self.flush()?;
+
+        Ok(())
+    }
+
+    fn recv_msg(&mut self) -> Result<Option<&[u8]>, Self::RecvError> {
+        loop {
+            match (self.recv_state, self.expected_state) {
+                //Stale Buffer state or recieved everything
+                (RxState::RxSize(x), RxState::RxSize(y))
+                | (RxState::RxBuffer(x), RxState::RxBuffer(y))
+                    if x == y =>
+                {
+                    match self.recv_state {
+                        RxState::RxSize(s) => {
+                            let len: &[u8; LEN_SIZE] = self.recv_buf[0..s].try_into().unwrap();
+                            let len: usize = u64::from_le_bytes(*len) as usize;
+
+                            ensure!(
+                                len <= msgs::RESPONSE_MSG_BUFFER_SIZE,
+                                "Oversized buffer ({len}) in psk buffer response."
+                            );
+
+                            self.recv_state = RxState::RxBuffer(0);
+                            self.expected_state = RxState::RxBuffer(len);
+                            continue;
+                        }
+                        RxState::RxBuffer(s) => {
+                            self.recv_state = RxState::RxSize(0);
+                            self.recv_state = RxState::RxSize(LEN_SIZE);
```

Comment:
Do you intent to assign `recv_state` twice? I suspect the second assignment is meant for `expected_state`.

Reply:
Oops, didn't intend that. Thanks, good catch!